src/java/picard/analysis/CollectWgsMetricsFromQuerySorted.java

```diff
@@ -139,38 +116,110 @@ protected int doWork() {
             pairToAnalyze = QuerySortedReadPairIteratorUtil.getNextReadPair(iterator);
         }
 
-        // finalize metrics
-        setUnusedMetrics(metrics);
-        metrics.GENOME_TERRITORY = reader.getFileHeader().getSequenceDictionary().getReferenceLength();
-        metrics.MEAN_COVERAGE = metrics.TOTAL_USABLE_BASES / (double)metrics.GENOME_TERRITORY;
-        metrics.PCT_EXC_DUPE = basesExcludedByDupes / (double)metrics.TOTAL_BASES;
-        metrics.PCT_EXC_MAPQ = basesExcludedByMapq / (double)metrics.TOTAL_BASES;
-        metrics.PCT_EXC_UNPAIRED = basesExcludedByPairing / (double)metrics.TOTAL_BASES;
-        metrics.PCT_EXC_BASEQ = basesExcludedByBaseq / (double)metrics.TOTAL_BASES;
-        metrics.PCT_EXC_OVERLAP = basesExcludedByOverlap / (double)metrics.TOTAL_BASES;
-        final double totalExcludedBases = metrics.TOTAL_BASES - metrics.TOTAL_USABLE_BASES;
-        metrics.PCT_EXC_TOTAL = totalExcludedBases / metrics.TOTAL_BASES;
-        metrics.MEAN_INSERT_SIZE = insertSizeSum / metrics.TOTAL_ORIENTED_PAIRS;
+        // finalize and write the metrics
+        final long genomeTerritory = reader.getFileHeader().getSequenceDictionary().getReferenceLength();
+        usableMetrics.metrics.GENOME_TERRITORY = genomeTerritory;
+        finalizeMetrics(usableMetrics);
+        rawMetrics.metrics.GENOME_TERRITORY = genomeTerritory;
+        finalizeMetrics(rawMetrics);
 
         final MetricsFile<QuerySortedSeqMetrics, Integer> out = getMetricsFile();
-        out.addMetric(metrics);
+        out.addMetric(usableMetrics.metrics);
+        out.addMetric(rawMetrics.metrics);
         out.write(OUTPUT);
-
         return 0;
     }
 
     /**
+     * Calculate the contribution to the intermediate metrics for a given read pair
+     *
+     * @param pairToAnalyze the read pair to grab metrics from
+     * @param metrics the intermediate metrics with all the data we need
+     * @param minimumMappingQuality the minimum mapping quality
+     * @param minimumBaseQuality the minimum base quality
+     */
+    private void calculateMetricsForRead(final QuerySortedReadPairIteratorUtil.ReadPair pairToAnalyze,
+                                         final IntermediateMetrics metrics,
+                                         final int minimumMappingQuality,
+                                         final int minimumBaseQuality) {
+
+        final boolean isProperPair = (pairToAnalyze.read2 != null);
```

Comment:
I was confused by the name of this variable since there is a similar flag field on a `SAMRecord`.  Perhaps you mean `hasSecondEnd` or `isUnpaired`?